<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; Estimating Effective Population Size and Key Stats – Analysing Genomic Data with dartRverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session03.html" rel="next">
<link href="./session01.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-a185852c63625fd9ffbdc57047c9a77e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-ef31e3f4d9d6eef55c54bdf6b9cd68ca.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">

<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>

<link href="site_libs/leaflet-1.3.1/leaflet.css" rel="stylesheet">

<script src="site_libs/leaflet-1.3.1/leaflet.js"></script>

<link href="site_libs/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">

<script src="site_libs/proj4-2.6.2/proj4.min.js"></script>

<script src="site_libs/Proj4Leaflet-1.0.1/proj4leaflet.js"></script>

<link href="site_libs/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">

<script src="site_libs/leaflet-binding-2.2.2/leaflet.js"></script>

<script src="site_libs/leaflet-providers-2.0.0/leaflet-providers_2.0.0.js"></script>

<script src="site_libs/leaflet-providers-plugin-2.2.2/leaflet-providers-plugin.js"></script>


  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session02.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Estimating Effective Population Size and Key Stats</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Analysing Genomic Data with dartRverse</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./speakers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Speaker Bios</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./breakk.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Importing and Handling Genomic Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Estimating Effective Population Size and Key Stats</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identifying Population Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Kinship, Pedigrees &amp; Mating Recommendations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assigning Individuals to Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">SNP Panel Selection &amp; Simulations for Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Simulations for Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Bonus! Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./feedback.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Feedback</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#calculating-the-number-of-alleles" id="toc-calculating-the-number-of-alleles" class="nav-link" data-scroll-target="#calculating-the-number-of-alleles">Calculating the number of Alleles</a>
  <ul>
  <li><a href="#number-of-alleles-na-for-snps" id="toc-number-of-alleles-na-for-snps" class="nav-link" data-scroll-target="#number-of-alleles-na-for-snps">Number of Alleles (Na) for SNPs:</a></li>
  <li><a href="#allelic-richness-ar-for-snps" id="toc-allelic-richness-ar-for-snps" class="nav-link" data-scroll-target="#allelic-richness-ar-for-snps">Allelic Richness (Ar) for SNPs:</a></li>
  </ul></li>
  <li><a href="#calculating-heterozygosity" id="toc-calculating-heterozygosity" class="nav-link" data-scroll-target="#calculating-heterozygosity">Calculating Heterozygosity</a>
  <ul>
  <li><a href="#observed-heterozygosity-ho" id="toc-observed-heterozygosity-ho" class="nav-link" data-scroll-target="#observed-heterozygosity-ho">Observed Heterozygosity (Ho):</a></li>
  <li><a href="#expected-heterozygosity-he" id="toc-expected-heterozygosity-he" class="nav-link" data-scroll-target="#expected-heterozygosity-he">Expected Heterozygosity (He)</a>
  <ul class="collapse">
  <li><a href="#decline-of-heterozygosity-over-time-in-an-ideal-population" id="toc-decline-of-heterozygosity-over-time-in-an-ideal-population" class="nav-link" data-scroll-target="#decline-of-heterozygosity-over-time-in-an-ideal-population">Decline of Heterozygosity Over Time in an Ideal Population</a></li>
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation"><strong>Interpretation</strong></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#hardy-weinberg-equilibrium-hwe" id="toc-hardy-weinberg-equilibrium-hwe" class="nav-link" data-scroll-target="#hardy-weinberg-equilibrium-hwe">Hardy-Weinberg Equilibrium (HWE)</a></li>
  <li><a href="#inbreeding-fis" id="toc-inbreeding-fis" class="nav-link" data-scroll-target="#inbreeding-fis">Inbreeding <em>F</em><sub>IS</sub></a></li>
  <li><a href="#fixed-and-private-alleles" id="toc-fixed-and-private-alleles" class="nav-link" data-scroll-target="#fixed-and-private-alleles">Fixed and private Alleles</a>
  <ul>
  <li><a href="#private-alleles" id="toc-private-alleles" class="nav-link" data-scroll-target="#private-alleles">Private Alleles</a></li>
  <li><a href="#fixed-alleles" id="toc-fixed-alleles" class="nav-link" data-scroll-target="#fixed-alleles">Fixed Alleles</a></li>
  </ul></li>
  <li><a href="#relatedness-and-kinship" id="toc-relatedness-and-kinship" class="nav-link" data-scroll-target="#relatedness-and-kinship">Relatedness and Kinship</a></li>
  <li><a href="#effective-population-size" id="toc-effective-population-size" class="nav-link" data-scroll-target="#effective-population-size">Effective Population Size</a>
  <ul>
  <li><a href="#current-effective-population-size" id="toc-current-effective-population-size" class="nav-link" data-scroll-target="#current-effective-population-size">Current effective population size</a></li>
  <li><a href="#historic-population-sizes" id="toc-historic-population-sizes" class="nav-link" data-scroll-target="#historic-population-sizes">Historic population sizes</a>
  <ul class="collapse">
  <li><a href="#cged-example" id="toc-cged-example" class="nav-link" data-scroll-target="#cged-example">CGED example</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#further-study" id="toc-further-study" class="nav-link" data-scroll-target="#further-study"><em>Further Study</em></a>
  <ul>
  <li><a href="#readings" id="toc-readings" class="nav-link" data-scroll-target="#readings">Readings</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Estimating Effective Population Size and Key Stats</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><em>Session Presenters</em></p>
<p><img src="images/Presenters2.png" class="img-fluid"></p>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>This session will cover the basic statistics that are used to study populations, mainly towards a conservation perspective. We will explore how to estimate effective population size (Ne), heterozygosity (Ho, He), and other key statistics using genomic data. The session will include practical exercises using R and the <code>dartRverse</code> packages.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/genetic_comp.jpg" class="img-fluid figure-img"></p>
<figcaption><a href="https://onlinelibrary.wiley.com/doi/10.1111/brv.12852">Hoban et al.&nbsp;2022</a></figcaption>
</figure>
</div>
<p>for this session you can use your own data, but feel free to use the example data provided in the <code>dartRverse</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>gls <span class="ot">&lt;-</span> possums.gl[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">31</span><span class="sc">:</span><span class="dv">35</span>),<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]  <span class="co">#small data set </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calculating-the-number-of-alleles" class="level2">
<h2 class="anchored" data-anchor-id="calculating-the-number-of-alleles">Calculating the number of Alleles</h2>
<p>In conservation genetics, the number of alleles (allelic richness) is important because it reflects the genetic diversity within a population. This diversity is crucial for:</p>
<ul>
<li><p>Adaptive potential – more alleles mean a greater chance that some individuals carry beneficial variants to cope with environmental changes or disease.</p></li>
<li><p>Long-term survival – populations with low allelic diversity are more vulnerable to inbreeding, genetic drift, and extinction.</p></li>
<li><p>Conservation decision-making – monitoring allele numbers helps identify populations that are genetically depauperate and may need management (e.g., genetic rescue).</p></li>
</ul>
<p>In short: more alleles = higher evolutionary potential.</p>
<p>There are two ways to meassure this quantity.</p>
<section id="number-of-alleles-na-for-snps" class="level3">
<h3 class="anchored" data-anchor-id="number-of-alleles-na-for-snps">Number of Alleles (Na) for SNPs:</h3>
<p>Since SNPs are biallelic by design, Na is either 1 or 2.</p>
<ul>
<li>If everyone has the same allele → Na = 1 (monomorphic)<br>
</li>
<li>If both alleles are present → Na = 2 (polymorphic)</li>
</ul>
</section>
<section id="allelic-richness-ar-for-snps" class="level3">
<h3 class="anchored" data-anchor-id="allelic-richness-ar-for-snps">Allelic Richness (Ar) for SNPs:</h3>
<p>Still measures how many alleles are present, but adjusted for sample size via rarefaction. In biallelic SNPs, Ar also ranges from 1 to 2, but:</p>
<ul>
<li>In small samples, rare alleles might be missed → Ar &lt; Na</li>
<li>Ar estimates the expected number of alleles if the sample had fewer individuals.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#by hand number of alleles</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as.matrix</span>(gls[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,])</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co">#count the number of alleles per locus in each population</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>nas1 <span class="ot">&lt;-</span> <span class="fu">gl.report.allelerich</span>(gls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>nas1<span class="sc">$</span><span class="st">`</span><span class="at">Allelic Richness per population</span><span class="st">`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  pop sum_corrected_richness mean_corrected_richness popsize      SD
1   A                     14                2.000000       5 0.00000
2   B                     12                1.714286       5 0.48795</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>nas2 <span class="ot">&lt;-</span> <span class="fu">gl.report.diversity</span>(gls, <span class="at">table =</span> <span class="st">"D"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.diversity 
  Processing genlight object with SNP data
Starting gl.filter.allna 
  Identifying and removing loci and individuals scored all 
                missing (NA)
  Deleting loci that are scored as all missing (NA)
  Deleting individuals that are scored as all missing (NA)
Completed: gl.filter.allna 
Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>

|   | nloci| m_0Da| sd_0Da| m_1Da| sd_1Da| m_2Da| sd_2Da|
|:--|-----:|-----:|------:|-----:|------:|-----:|------:|
|A  |     7| 2.000|  0.000| 1.921|  0.076| 1.860|  0.131|
|B  |     7| 1.714|  0.488| 1.641|  0.453| 1.595|  0.439|


pairwise non-missing loci

|   |  A|  B|
|:--|--:|--:|
|A  | NA| NA|
|B  |  7| NA|


0_D_beta

|   |     A|     B|
|:--|-----:|-----:|
|A  |    NA| 0.244|
|B  | 1.143|    NA|


1_D_beta

|   |     A|     B|
|:--|-----:|-----:|
|A  |    NA| 0.094|
|B  | 1.077|    NA|


2_D_beta

|   |    A|     B|
|:--|----:|-----:|
|A  |   NA| 0.198|
|B  | 1.16|    NA|
Completed: gl.report.diversity </code></pre>
</div>
</div>
<p>A nice function to have is to run a bootstrapped simulation that selects a random sample of individuals and calculates the allelic richness for that sample and compares it from a sample of the same number of individuals from the combined population.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span> <span class="fu">gl.report.nall</span>(gls, <span class="at">simlevels =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">reps =</span> <span class="dv">20</span>, <span class="at">ncores =</span> <span class="dv">1</span>)  <span class="co">#change the number of cores if you have more available</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.nall 
  Processing genlight object with SNP data
Starting gl.filter.allna 
  Identifying and removing loci that are all missing (NA) 
                    in any one population
  Deleting loci that are all missing (NA) in any one population
  Warning: no loci listed to delete! Genlight object returned unchanged
Completed: gl.filter.allna 
Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.report.nall </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class img-fluid" width="48"> Rerun the analysis with your own data (or use the glb dataset provided here)</p>
</div>
</div>
</section>
</section>
<section id="calculating-heterozygosity" class="level2">
<h2 class="anchored" data-anchor-id="calculating-heterozygosity">Calculating Heterozygosity</h2>
<section id="observed-heterozygosity-ho" class="level3">
<h3 class="anchored" data-anchor-id="observed-heterozygosity-ho">Observed Heterozygosity (Ho):</h3>
<p>Observed heterozygosity (Ho) is the proportion of individuals in a population that are heterozygous at a given locus. It is calculated as per individual and often averaged per population.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.heterozygosity</span>(gls, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating observed heterozygosity for individuals
  Note: No adjustment for invariant loci (n.invariant set to 0)
Starting gl.colors 
Selected color type 2 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code> ind.name        Ho f.hom.ref f.hom.alt n.Loc
        1 0.4285714 0.0000000 0.5714286     7
        2 0.8571429 0.1428571 0.0000000     7
        3 0.4285714 0.2857143 0.2857143     7
        4 0.2857143 0.5714286 0.1428571     7
        5 0.4285714 0.0000000 0.5714286     7
       31 0.4285714 0.2857143 0.2857143     7
       32 0.7142857 0.1428571 0.1428571     7
       33 0.4285714 0.2857143 0.2857143     7
       34 0.2857143 0.5714286 0.1428571     7
       35 0.4285714 0.2857143 0.2857143     7
Completed: gl.report.heterozygosity </code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.heterozygosity</span>(gls, <span class="at">method =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  pop n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
A   A     5     7         1       7       0         0 0.485714 0.157359
B   B     5     7         1       5       2         0 0.457143 0.359894
      HoSE   He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS   FISSD
A 0.059476 0.46 0.038297 0.014475 0.511111 0.042552 0.016083  0.02898 0.36056
B 0.136027 0.32 0.226274 0.085524 0.355556 0.251416 0.095026 -0.27500 0.33541
     FISSE
A 0.136279
B       NA
Completed: gl.report.heterozygosity </code></pre>
</div>
</div>
<p><strong>There is a lot of discussion on the best way to calculate heterozygosity</strong>, but the most common method is to use the proportion of heterozygous individuals at each locus. This is “okay” if you compare individuals and populations of the same species (but see Sopniewski and Catullo 2022 and Schmidt et al.&nbsp;2021 for a discussion on the limitations of this approach). The main problem is that when filtering for low quality (low read depth) often heterozygous loci are lost, which can bias the results. In addition to calculate a comparable heterozygosity across species, we would need to know the number of invariant sites, which are not easily obtained. dartR aims to estimate this number from closely neighbouring loci, if you are using dartR data (secondaries). This estimate is based on a Poisson distribution assumption which is most likely not a good idea as it underestimates the number of invariant sites. We are currently aiming to implement a better method to estimate the number of invariant sites, but for now we will use the <code>gl.report.heterozygosity</code> function to calculate Ho. Schmidt et al.&nbsp;(2021) suggests to calculate genome-wide/autosomal heterozygosity, but it basically means you need whole genome sequences, which is often not available. In summary be careful to calculate Heterozygosity, especially when comparing across species or uneven sampled populations. A rarefaction approach is often used to standardise the sample size. We can do that with the <code>gl.report.heterozygosity</code> function, which allows us to calculate Ho for each individual and then average it per population in combination of <code>gl.subsample.ind()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Create a sample data sets </span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (two populations, once with 10 individuals and one with 5 individuals)</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>gls1 <span class="ot">&lt;-</span> possums.gl[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">31</span><span class="sc">:</span><span class="dv">35</span>),]</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co">#cr</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>subfun <span class="ot">&lt;-</span> <span class="cf">function</span>(x) { </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  xx <span class="ot">&lt;-</span> <span class="fu">gl.subsample.ind</span>(x, <span class="at">n =</span> <span class="dv">5</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(xx, <span class="at">method =</span> <span class="st">"pop"</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(out<span class="sc">$</span>Ho)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="fu">subfun</span>(gls1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Warning: Input genlight objects both lack individual metrics</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.343 0.436</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subfun</span>(gls1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Warning: Input genlight objects both lack individual metrics</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.366 0.433</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">#ignore the warnings</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, <span class="cf">function</span>(x) <span class="fu">subfun</span>(gls1)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(res) <span class="ot">&lt;-</span> <span class="fu">popNames</span>(gls1)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(<span class="fu">t</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">t</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       A                B         
 Min.   :0.3060   Min.   :0.4240  
 1st Qu.:0.3360   1st Qu.:0.4350  
 Median :0.3450   Median :0.4380  
 Mean   :0.3457   Mean   :0.4376  
 3rd Qu.:0.3590   3rd Qu.:0.4410  
 Max.   :0.3780   Max.   :0.4440  </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Task
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class img-fluid" width="48"> Try with your own data set</p>
</div>
</div>
</section>
<section id="expected-heterozygosity-he" class="level3">
<h3 class="anchored" data-anchor-id="expected-heterozygosity-he">Expected Heterozygosity (He)</h3>
<p>Expected heterozygosity (He) is a key measure of genetic diversity in conservation genetics, representing the probability that two alleles randomly drawn from a population are different. High He indicates a genetically diverse population, which is critical for adaptive potential and long-term viability. In contrast, low He can signal inbreeding, genetic drift, or population bottlenecks. Monitoring He helps conservationists assess population health, guide management actions such as translocations or genetic rescue, and evaluate the success of captive breeding programs in maintaining genetic variation..</p>
<section id="decline-of-heterozygosity-over-time-in-an-ideal-population" class="level4">
<h4 class="anchored" data-anchor-id="decline-of-heterozygosity-over-time-in-an-ideal-population">Decline of Heterozygosity Over Time in an Ideal Population</h4>
<p>In an <strong>ideal population</strong>, the expected heterozygosity (He) declines over time due to <strong>genetic drift</strong>, even in the absence of selection, mutation, or migration. The rate of this decline is governed by the <strong>effective population size (Ne)</strong> and follows this mathematical law:</p>
<p><span class="math display">\[
H_t = H_0 \left(1 - \frac{1}{2N_e} \right)^t
\]</span></p>
<ul>
<li>( H_0 ): initial heterozygosity<br>
</li>
<li>( H_t ): heterozygosity after ( t ) generations<br>
</li>
<li>( N_e ): effective population size<br>
</li>
<li>( (1 - ) ): per-generation retention of heterozygosity</li>
</ul>
</section>
<section id="interpretation" class="level4">
<h4 class="anchored" data-anchor-id="interpretation"><strong>Interpretation</strong></h4>
<ul>
<li>Larger ( N_e ) → slower loss of He<br>
</li>
<li>Small ( N_e ) → rapid loss of He due to drift</li>
</ul>
<p>This law underscores why <strong>maintaining a large Ne</strong> is a central goal in conservation: to preserve genetic variation over time and reduce the risk of inbreeding and loss of adaptive potential.</p>
<p>Expected heterozygosity (He) is a key measure of genetic diversity in conservation genetics, representing the probability that two alleles randomly drawn from a population are different. High He indicates a genetically diverse population, which is critical for adaptive potential and long-term viability. In contrast, low He can signal inbreeding, genetic drift, or population bottlenecks. Monitoring He helps conservationists assess population health, guide management actions such as translocations or genetic rescue, and evaluate the success of captive breeding programs in maintaining genetic variation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(gls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>gg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  pop n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
A   A     5     7         1       7       0         0 0.485714 0.157359
B   B     5     7         1       5       2         0 0.457143 0.359894
      HoSE   He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS   FISSD
A 0.059476 0.46 0.038297 0.014475 0.511111 0.042552 0.016083  0.02898 0.36056
B 0.136027 0.32 0.226274 0.085524 0.355556 0.251416 0.095026 -0.27500 0.33541
     FISSE
A 0.136279
B       NA</code></pre>
</div>
</div>
<p>Expected heterozygosity can be standardised by sample size (because allele frequencies are estimated and they are missing rare alleles at low sample size, hence is biased downwards. Therefore the correction is 2n/(2n-1) where n is the number of individuals. This is then called uHe, but is mainly important due to low sample sizes. Again you can use a rarefaction approach to standardise the sample size, which is implemented in the <code>gl.report.heterozygosity</code> function.</p>
</section>
</section>
</section>
<section id="hardy-weinberg-equilibrium-hwe" class="level2">
<h2 class="anchored" data-anchor-id="hardy-weinberg-equilibrium-hwe">Hardy-Weinberg Equilibrium (HWE)</h2>
<p>HWE is a principle that describes the genetic variation in a population under certain conditions. It states that allele and genotype frequencies will remain constant from generation to generation in the absence of evolutionary influences.</p>
<p>In conservation genetics, testing for HWE is important because:</p>
<ul>
<li>Deviations from HWE can indicate inbreeding, genetic drift, or selection pressures.</li>
<li>Helps identify populations at risk of losing genetic diversity.</li>
<li>Can inform management strategies to maintain genetic health.</li>
</ul>
<p>To test for HWE, we can use the <code>gl.report.hwe</code> function, which performs a chi-squared test for each locus and returns the p-values.</p>
<p>This can be done on population level (subset=“each”) or on the whole dataset (subset=“all”).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#does not make much sense (sample size too low)</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.hwe</span>(gls,<span class="at">subset =</span> <span class="st">"each"</span>,<span class="at">min_sample_size =</span> <span class="dv">1</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 methods overwritten by 'ggtern':
  method           from   
  grid.draw.ggplot ggplot2
  plot.ggplot      ggplot2
  print.ggplot     ggplot2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?
`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># only 10 loci... Lets try with more</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.hwe</span>(possums.gl[<span class="dv">1</span><span class="sc">:</span><span class="dv">90</span>,],<span class="at">subset =</span> <span class="st">"each"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.hwe 
  Processing genlight object with SNP data
  Analysing each population separately
Starting gl.colors 
Selected color type 2c 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    Reporting significant departures from Hardy-Weinberg 
            Equilibrium
    NB: Departures significant at the alpha level of 0.05 are listed
    Adjustment of p-values for multiple comparisons vary
                    with sample size
 Population  Locus Hom_1   Het Hom_2     N         Prob    Sig Prob.adj Sig.adj
     &lt;char&gt; &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;num&gt;        &lt;num&gt; &lt;char&gt;   &lt;lgcl&gt;  &lt;lgcl&gt;
          B   X114    19     7     4    30 0.0481152745    sig       NA      NA
          B   X120    22     5     3    30 0.0322894276    sig       NA      NA
          B   X123     3    24     3    30 0.0026880391    sig       NA      NA
          C   X128    10    20     0    30 0.0108195955    sig       NA      NA
          C   X140     4    22     4    30 0.0261558384    sig       NA      NA
          B   X142     2    21     7    30 0.0275978824    sig       NA      NA
          A   X152    18     7     5    30 0.0248050366    sig       NA      NA
          B   X171    22     5     3    30 0.0322894276    sig       NA      NA
          B   X177     1    24     5    30 0.0009308178    sig       NA      NA
          A   X181     4     7    19    30 0.0481152745    sig       NA      NA
          B   X198     9     9    12    30 0.0324438457    sig       NA      NA
          A    X30    11    19     0    30 0.0279783750    sig       NA      NA
          B    X31     3    22     5    30 0.0250646233    sig       NA      NA
          B    X47     6    22     2    30 0.0118328520    sig       NA      NA
          A    X48     3    22     5    30 0.0250646233    sig       NA      NA
          A    X50     5    23     2    30 0.0079779773    sig       NA      NA
          B    X55    11    19     0    30 0.0279783750    sig       NA      NA
          B    X65     8     7    15    30 0.0066040187    sig       NA      NA
  npop
 &lt;int&gt;
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1
     1

Completed: gl.report.hwe </code></pre>
</div>
</div>
</section>
<section id="inbreeding-fis" class="level2">
<h2 class="anchored" data-anchor-id="inbreeding-fis">Inbreeding <em>F</em><sub>IS</sub></h2>
<p>Inbreeding coefficient (<em>F</em><sub>IS</sub>) is a measure of the degree of inbreeding in a population. It quantifies the reduction in heterozygosity due to inbreeding compared to a randomly mating population. A positive <em>F</em><sub>IS</sub> indicates inbreeding, while a negative value suggests outbreeding or excess heterozygosity.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.heterozygosity</span>(gls)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  pop n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
A   A     5     7         1       7       0         0 0.485714 0.157359
B   B     5     7         1       5       2         0 0.457143 0.359894
      HoSE   He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS   FISSD
A 0.059476 0.46 0.038297 0.014475 0.511111 0.042552 0.016083  0.02898 0.36056
B 0.136027 0.32 0.226274 0.085524 0.355556 0.251416 0.095026 -0.27500 0.33541
     FISSE
A 0.136279
B       NA
Completed: gl.report.heterozygosity </code></pre>
</div>
</div>
</section>
<section id="fixed-and-private-alleles" class="level2">
<h2 class="anchored" data-anchor-id="fixed-and-private-alleles">Fixed and private Alleles</h2>
<p>Private alleles are alleles found only in a single population, while fixed alleles are alleles that occur at 100% frequency within a population. In conservation genetics, private alleles can indicate unique evolutionary history or local adaptation and are useful for identifying distinct populations or units for conservation. Fixed alleles, on the other hand, may signal a loss of genetic diversity due to drift or inbreeding. Monitoring both helps assess population structure, track gene flow, and guide decisions about mixing or isolating populations. Private alleles can also be used to identify assymetry in geneflow (Campbell et al.&nbsp;2021).</p>
<section id="private-alleles" class="level3">
<h3 class="anchored" data-anchor-id="private-alleles">Private Alleles</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.map.interactive</span>(possums.gl[<span class="dv">1</span><span class="sc">:</span><span class="dv">120</span>,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.map.interactive 
  Processing genlight object with SNP data
Completed: gl.map.interactive </code></pre>
</div>
<div class="cell-output-display">
<div class="leaflet html-widget html-fill-item" id="htmlwidget-958a4bcb6c99b27c48b6" style="width:100%;height:464px;"></div>
<script type="application/json" data-for="htmlwidget-958a4bcb6c99b27c48b6">{"x":{"options":{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}}},"calls":[{"method":"addTiles","args":["https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",null,null,{"minZoom":0,"maxZoom":18,"tileSize":256,"subdomains":"abc","errorTileUrl":"","tms":false,"noWrap":false,"zoomOffset":0,"zoomReverse":false,"opacity":1,"zIndex":1,"detectRetina":false,"attribution":"&copy; <a href=\"https://openstreetmap.org/copyright/\">OpenStreetMap<\/a>,  <a href=\"https://opendatacommons.org/licenses/odbl/\">ODbL<\/a>"}]},{"method":"addCircles","args":[[-35.06726602006179,-35.06955064191888,-35.07673446942515,-35.06253077174035,-35.06596335361223,-35.07240653278109,-35.06945030472441,-35.05851152550337,-35.05985319383298,-35.06269461665501,-35.07283232198104,-35.06785969168114,-35.06140452612459,-35.07000849226291,-35.06262776758465,-35.0700717233947,-35.06763775026968,-35.06326203785833,-35.07286252370781,-35.05929856231865,-35.06097874939351,-35.06590367721348,-35.05911385039392,-35.06511327075684,-35.06618017560768,-35.0597295335847,-35.06424845372028,-35.06585027391014,-35.06321742421637,-35.07252226008766,-35.46055211706051,-35.45161364369143,-35.45902228106662,-35.44643386742302,-35.46243693185566,-35.45417959421039,-35.45018417344416,-35.45924311403174,-35.44890760005696,-35.4521796574947,-35.45988605372478,-35.45539809735926,-35.45501903103015,-35.4586703397519,-35.45686504707611,-35.46479149128798,-35.45078857294831,-35.45815868254312,-35.45466795585465,-35.45039956768484,-35.46205083650445,-35.4598597488244,-35.45619141886416,-35.4635740328022,-35.4552718590738,-35.45678531683682,-35.45158548805967,-35.46508625696213,-35.45351399235329,-35.45707818180141,-35.4305411459387,-35.42685899024227,-35.42643498625106,-35.42482238917844,-35.42153397361244,-35.42609105710914,-35.42688871882999,-35.42255791970729,-35.42931971756748,-35.42770966763658,-35.42101151611572,-35.42404413631208,-35.43260612963641,-35.42462356150683,-35.43076173989826,-35.4333022917218,-35.4283444715299,-35.43703873162963,-35.42927860409449,-35.425665365613,-35.42956965923308,-35.41644235863421,-35.42824801358195,-35.43278182931387,-35.42581793190112,-35.43272295602974,-35.42391325313334,-35.43013284624325,-35.42606939645366,-35.43503345463999,-35.32169712970318,-35.32487649457556,-35.32495027923733,-35.33010001632034,-35.32848842711304,-35.3353022348978,-35.32360991543328,-35.32751590778332,-35.33408666786472,-35.33098529253679,-35.32566957462272,-35.33030282665679,-35.32736989170276,-35.33371917323233,-35.33427961601613,-35.32578978209212,-35.32223144080267,-35.32890169955976,-35.33132488967286,-35.33897980880262,-35.32566382444671,-35.33328290822069,-35.32046076559659,-35.33062493966311,-35.3265500114555,-35.33813236025612,-35.33011564763665,-35.31420499230119,-35.33712599234254,-35.32168066764579],[149.2869862369311,149.2746309785811,149.2722429154469,149.2812307278812,149.2750973419055,149.2781025173445,149.2788742249669,149.2860977673041,149.2932414893118,149.2755740248056,149.2836017973049,149.2668731448335,149.2834512457887,149.2751440245323,149.2874537547859,149.2808386423436,149.2931575381382,149.2800710287609,149.2786355442883,149.2725685188075,149.2777926804528,149.2841536843803,149.286859204825,149.2756010351737,149.2798786167938,149.2803295265397,149.2749598709062,149.274843528577,149.2798726878657,149.2833821406627,149.0288179804404,149.0277180830705,149.0312045633892,149.0376413352734,149.020213771592,149.031464519822,149.0263965249884,149.0334447890848,149.0319666754673,149.0267593928923,149.0246837668797,149.0305891411105,149.0333227917825,149.0300755275159,149.0274354086494,149.0256585381179,149.0270384143907,149.0356977904418,149.0268846492132,149.0368917779229,149.0228416033965,149.0324760640069,149.0282992350292,149.0314982386245,149.0360673401843,149.0207172222744,149.0400934578165,149.0233656493286,149.0341314983903,149.0283495735485,149.0574687540712,149.0642340451419,149.0677556332095,149.054291984695,149.0602994007142,149.0601359082545,149.0628131330281,149.0535954115097,149.058145221787,149.0569703969507,149.0554037144645,149.0595432368244,149.0596963628989,149.0647007349419,149.0597112056045,149.0626633884261,149.0547451500818,149.05438612097,149.0635747895413,149.0658366566245,149.06260850334,149.0572024129704,149.0577147431848,149.0545997458017,149.0626555521874,149.0533198080399,149.0632658470781,149.0584276835635,149.0592254429095,149.0664611869001,149.3407810789724,149.3443086887442,149.345574440209,149.3294786770262,149.3432051441674,149.3387756642478,149.3385776355366,149.3318337391863,149.3357765252715,149.3425597241434,149.3345677489617,149.3461888279114,149.333187906449,149.3404463687952,149.3384208988126,149.3380757594117,149.334741497081,149.337551214032,149.3371697887127,149.3375770393958,149.3465796576442,149.3371321827274,149.3361313733821,149.3415461243541,149.3430198145683,149.3401600929371,149.3385720295742,149.3343231627318,149.3345616070885,149.3359777747675],10,null,null,{"interactive":true,"className":"","stroke":true,"color":["#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF"],"weight":10,"opacity":0.8,"fill":true,"fillColor":["#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#FF0000","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#80FF00","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#00FFFF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF","#8000FF"],"fillOpacity":0.2},["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120"],null,null,{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null,null]},{"method":"addMarkers","args":[[-35.06585614987745,-35.45634649838929,-35.42767222710986,-35.32860077260636],[149.280051548008,149.0297248441548,149.0597150725238,149.3385600728948],{"iconUrl":{"data":"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR4nGP6zwAAAgcBApocMXEAAAAASUVORK5CYII=","index":0},"iconWidth":1,"iconHeight":1},null,null,{"interactive":true,"draggable":false,"keyboard":true,"title":"","alt":"","zIndexOffset":0,"opacity":1,"riseOnHover":false,"riseOffset":250},null,null,null,null,["A","B","C","D"],{"interactive":false,"permanent":true,"direction":"top","opacity":1,"offset":[0,0],"textsize":"12px","textOnly":true,"className":"","sticky":true},null]},{"method":"addProviderTiles","args":["Esri.NatGeoWorldMap",null,null,{"errorTileUrl":"","noWrap":false,"detectRetina":false}]}],"limits":{"lat":[-35.46508625696213,-35.05851152550337],"lng":[149.020213771592,149.3465796576442]}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.pa</span>(possums.gl[<span class="dv">1</span><span class="sc">:</span><span class="dv">120</span>,] )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.pa 
  Processing genlight object with SNP data
  Warning: no loci listed to keep! Genlight object returned unchanged
  Warning: no loci listed to keep! Genlight object returned unchanged
  p1 p2 pop1 pop2 N1 N2 fixed priv1 priv2 Chao1 Chao2 totalpriv   AFD asym
1  1  2    A    B 30 30     0     1    24    NA     0        25 0.309   NA
2  1  3    A    C 30 30     0     1    24    NA     0        25 0.302   NA
3  1  4    A    D 30 30     2    49    21     0     0        70 0.370   NA
4  2  3    B    C 30 30     0     0     0     0     0         0 0.180   NA
5  2  4    B    D 30 30     1    52     1     0    NA        53 0.332   NA
6  3  4    C    D 30 30     1    52     1     0    NA        53 0.314   NA
  asym.sig
1       NA
2       NA
3       NA
4       NA
5       NA
6       NA
  Table of private alleles and fixed differences returned
Completed: gl.report.pa </code></pre>
</div>
</div>
</section>
<section id="fixed-alleles" class="level3">
<h3 class="anchored" data-anchor-id="fixed-alleles">Fixed Alleles</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.fixed.diff</span>(possums.gl[<span class="dv">1</span><span class="sc">:</span><span class="dv">120</span>,])  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.fixed.diff 
  Processing genlight object with SNP data
  Comparing populations for absolute fixed differences
  Monomorphic loci removed
  Comparing populations pairwise -- this may take time. Please be patient
Completed: gl.fixed.diff </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$gl
 ********************
 *** DARTR OBJECT ***
 ********************

 ** 120 genotypes,  200 SNPs , size: 233.8 Kb

    missing data: 0 (=0 %) scored as NA

 ** Genetic data
   @gen: list of 120 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  120 individual labels
   @loc.names:  200 locus labels
   @loc.all:   no allele labels
   @pop: population of each individual (group size range: 30-30)
   @other: a list containing: xy, loc.metrics, loc.metrics.flags, verbose, history, latlon 
   @other$latlon[g]: coordinates for all individuals are attached

$fd
  A B C
B 0    
C 0 0  
D 2 1 1

$pcfd
  A B C
B 0    
C 0 0  
D 1 0 0

$nobs
   A  B  C  D
A NA 60 60 60
B 60 NA 60 60
C 60 60 NA 60
D 60 60 60 NA

$nloc
    A   B   C   D
A  NA 200 200 200
B 200  NA 200 200
C 200 200  NA 200
D 200 200 200  NA

$expfpos
     [,1] [,2] [,3] [,4]
[1,]   NA   NA   NA   NA
[2,]   NA   NA   NA   NA
[3,]   NA   NA   NA   NA
[4,]   NA   NA   NA   NA

$sdfpos
     [,1] [,2] [,3] [,4]
[1,]   NA   NA   NA   NA
[2,]   NA   NA   NA   NA
[3,]   NA   NA   NA   NA
[4,]   NA   NA   NA   NA

$pval
     [,1] [,2] [,3] [,4]
[1,]   NA   NA   NA   NA
[2,]   NA   NA   NA   NA
[3,]   NA   NA   NA   NA
[4,]   NA   NA   NA   NA

attr(,"class")
[1] "fd"</code></pre>
</div>
</div>
</section>
</section>
<section id="relatedness-and-kinship" class="level2">
<h2 class="anchored" data-anchor-id="relatedness-and-kinship">Relatedness and Kinship</h2>
<p>Relatedness and kinship are measures of genetic relatedness between individuals in a population. They are crucial for understanding population structure, mating systems, and the potential for inbreeding.</p>
<p>Relatedness in dartRverse is calculated via the <code>gl.grm</code> function:</p>
<p>Based on the A.mat function (package rrBLUP) R estimates a genomic additive relationship matrix (GRM) based on SNP genotype data. The matrix it produces approximates realized additive genetic relationships between individuals, capturing shared alleles weighted by allele frequencies. The additive relationship matrix (often denoted AA) quantifies the expected proportion of alleles shared identical-by-descent (IBD) between individuals due to additive genetic effects. This generates an n×n matrix where each entry reflects the genomic similarity between two individuals.</p>
<p>Each entry <span class="math inline">\(A_ij\)</span> reflects the expected genetic relatedness between individuals i and j under the additive genetic model. Diagonal entries <span class="math inline">\(A_ii\)</span> represent the self-relatedness or inbreeding coefficient (1+Fi). Offf-diagonals <span class="math inline">\(A_ij\)</span> represent the proportion of alleles shared IBD between individuals. It produces values centered around 0 (unrelated) and 1 (identical).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">#simulate some relatedness data</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>glsim <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">50</span>, <span class="at">nlocs =</span> <span class="dv">1000</span>)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>Amat <span class="ot">&lt;-</span> <span class="fu">gl.grm</span>(glsim)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.grm 
  Processing genlight object with SNP data
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.grm </code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="do">### F is diag(A)-1</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># centered around 1</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">diag</span>(Amat)<span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">#proportion of shared alleles  IBD between individuals</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>relA <span class="ot">&lt;-</span> Amat</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(relA)<span class="ot">&lt;-</span><span class="cn">NA</span> </span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(relA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-17-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>So we have created a population of <strong>50 individuals</strong> with <strong>1000 loci</strong>, and calculated the Amat. The diagonal entries represent the inbreeding coefficient (1+Fi), so Fi is centered around 0, indicating that individuals are not highly inbred. The off-diagonal entries represent the proportion of alleles shared IBD between individuals, which is centered around 0, indicating that individuals are not closely related. As we have simulated theindividuals, we know that they are not related, so this is expected.</p>
<p>Now lets add some offsprings to the mix and see how the turn up in the A mat.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create an 10 offsprings from indivivdual 1 and 2</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>offsprings <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(glsim[<span class="dv">1</span>,], glsim[<span class="dv">2</span>,], <span class="at">noffpermother =</span> <span class="dv">10</span>, <span class="at">popname =</span> <span class="st">"off12"</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>glsimoff <span class="ot">&lt;-</span> <span class="fu">rbind</span>(glsim, offsprings)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>Amat <span class="ot">&lt;-</span> <span class="fu">gl.grm</span>(glsimoff)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.grm 
  Processing genlight object with SNP data
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.grm </code></pre>
</div>
</div>
<p>And here you can see the offsprings in the A mat, they are related to the parents (1 and 2) and to each other, but not to the rest of the population. There relatedness value (the proportion of alleles that individuals share IBD at any locus) is 0.5 (the expected value for half siblings and offsprings).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># matrix between offsprings and parents</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>Apo <span class="ot">&lt;-</span> Amat[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">51</span><span class="sc">:</span><span class="dv">60</span>),<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">51</span><span class="sc">:</span><span class="dv">60</span>) ]</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(Apo,<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          1     2 Po_1 Po_2 Po_3 Po_4 Po_5 Po_6 Po_7 Po_8 Po_9 Po_10
1      0.83 -0.13 0.32 0.31 0.38 0.34 0.39 0.35 0.32 0.31 0.32  0.34
2     -0.13  0.83 0.40 0.31 0.31 0.29 0.37 0.35 0.33 0.32 0.30  0.36
Po_1   0.32  0.40 0.93 0.31 0.37 0.30 0.45 0.42 0.38 0.36 0.29  0.37
Po_2   0.31  0.31 0.31 0.81 0.25 0.29 0.32 0.30 0.30 0.24 0.32  0.41
Po_3   0.38  0.31 0.37 0.25 0.93 0.31 0.40 0.30 0.35 0.22 0.35  0.31
Po_4   0.34  0.29 0.30 0.29 0.31 0.82 0.29 0.35 0.35 0.24 0.32  0.37
Po_5   0.39  0.37 0.45 0.32 0.40 0.29 0.95 0.35 0.31 0.43 0.24  0.38
Po_6   0.35  0.35 0.42 0.30 0.30 0.35 0.35 0.86 0.39 0.28 0.35  0.36
Po_7   0.32  0.33 0.38 0.30 0.35 0.35 0.31 0.39 0.83 0.27 0.33  0.36
Po_8   0.31  0.32 0.36 0.24 0.22 0.24 0.43 0.28 0.27 0.86 0.31  0.32
Po_9   0.32  0.30 0.29 0.32 0.35 0.32 0.24 0.35 0.33 0.31 0.83  0.28
Po_10  0.34  0.36 0.37 0.41 0.31 0.37 0.38 0.36 0.36 0.32 0.28  0.90</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="do">## only related to parents and each other</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>relpo <span class="ot">&lt;-</span> Apo</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(relpo)<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(relpo, <span class="at">main =</span> <span class="st">"Relatedness between offsprings and parents"</span>, <span class="at">xlab =</span> <span class="st">"Relatedness"</span>, <span class="at">ylab =</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Now lets look at the kinship matrix, which is calculated via the <code>gl.run.EMIBD9</code> function. For this function to work you need to have the <code>emibd9</code> package installed. An easy way to do so, is using the <code>gl.download.binary</code> in the dartRverse package. We asked for permission from the Author ()</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install emibd9 from Wang 2022 in the temporary directory</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>dir_emibd <span class="ot">&lt;-</span> <span class="st">'./binaries/'</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>dir <span class="ot">&lt;-</span> <span class="fu">gl.download.binary</span>(<span class="st">"emibd9"</span>, <span class="at">out.dir =</span> dir_emibd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Downloaded binary to  C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq\file124410f6598 
Unzipped binary to ./binaries//emibd9</code></pre>
</div>
</div>
<p>When estimating relatedness and kinship, there is <strong>one major difficulty</strong>. We need to have a good estimate of the <strong>allele frequencies</strong> in the population. If we have a population with a lot of missing data, or if we have a population with a lot of inbred individuals the estimates of relatedness and kinship will be biased. In the best case you would estimate allele frequencies from a sample of unrelated individuals and then use these allele frequencies to estimate relatedness and kinship in the whole population. However, this is rarely possible, especially in conservation genetics where we often work with small populations and have only one sample</p>
<p><strong>EMIBD9</strong> is aiming to take care of that, but using an algorithm that aims to estimate allele frequencies and kinship in a ‘shinkage’ fashion so both at the same time and optimising the result. It performs really well in simulations. Please note there are many other relatedness estimates, but they all suffer from this problem and often give negative results, which in principle should not happen. Lets use EMIBD9 to estimate the kinship matrix for our simulated population.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co">#ignore the warnings...</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>kinmat <span class="ot">&lt;-</span> <span class="fu">gl.run.EMIBD9</span>(glsimoff, <span class="at">Inbreed =</span> <span class="dv">1</span>, <span class="at">emibd9.path =</span> dir_emibd)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.run.EMIBD9 
  Processing genlight object with SNP data
  Found necessary files to run EMIBD9.
Exporting individual diversity and inbreeding values 
  Returning a list containing the input gl object, a square matrix  of pairwise kinship, and the raw EMIBD9 results table as follows:
           $rel -- a square matrix of relatedness 
           $raw -- raw EMIBD9 results table 
           $processed -- EMIBD9 results without self and redundant comparison 
           $inbreeding -- Individual diversity and inbreeding (if requested) 
Starting gl.colors 
Selected color type div 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'dendextend':
  method     from 
  rev.hclust vegan</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.run.EMIBD9 </code></pre>
</div>
</div>
<p>As before we find the related individuals. Lets check the results as before:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="do">### F is diag(kinmat$rel)-1</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>((<span class="dv">2</span><span class="sc">*</span><span class="fu">diag</span>(kinmat<span class="sc">$</span>rel) <span class="sc">-</span> <span class="dv">1</span>), <span class="at">main =</span> <span class="st">"Kinship"</span>, <span class="at">xlab =</span> <span class="st">"Kinship"</span>, <span class="at">ylab =</span> <span class="st">"Frequency"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co">#kinship values between parents and offsprings and between offsprings =0.25</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>kinpo <span class="ot">&lt;-</span> kinmat<span class="sc">$</span>rel[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">51</span><span class="sc">:</span><span class="dv">60</span>),<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">51</span><span class="sc">:</span><span class="dv">60</span>) ]</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(kinpo)<span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(kinpo, <span class="at">main =</span> <span class="st">"Kinship between offsprings and parents"</span>, <span class="at">xlab =</span> <span class="st">"Kinship"</span>, <span class="at">ylab =</span> <span class="st">"Frequency"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-22-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">#variance is different. </span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Parent off spring variance is zero.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Finally lets compare the relatedness estimates using gl.grm and EMIBD9. The results are very similar, but in theory the EMIBD9 estimates should be more precise, as they take into account the non independence of allele frequencies in the population.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co">#compare the two matrices</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>Arel <span class="ot">&lt;-</span> Amat[<span class="fu">lower.tri</span>(Amat, <span class="at">diag =</span> <span class="cn">FALSE</span>)]</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>kinrel <span class="ot">&lt;-</span> kinmat<span class="sc">$</span>rel[<span class="fu">lower.tri</span>(kinmat<span class="sc">$</span>rel, <span class="at">diag =</span> <span class="cn">FALSE</span>)]</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(Arel, kinrel, </span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Relatedness (gl.grm)"</span>, </span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Kinship (EMIBD9)"</span>, </span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Comparison of F values from gl.grm and EMIBD9"</span>,</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="effective-population-size" class="level2">
<h2 class="anchored" data-anchor-id="effective-population-size">Effective Population Size</h2>
<section id="current-effective-population-size" class="level3">
<h3 class="anchored" data-anchor-id="current-effective-population-size">Current effective population size</h3>
<p>Effective population size (Ne) is the size of an idealized population that would experience genetic drift or inbreeding at the same rate as the observed population.</p>
<p>It is almost always smaller than the actual census size (N) due to factors like unequal sex ratios, variation in reproductive success, or population size fluctuations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install the Neestimator package</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>dir <span class="ot">&lt;-</span> dartRverse<span class="sc">::</span><span class="fu">gl.download.binary</span>(<span class="st">"neestimator"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Downloaded binary to  C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq\file124431876020 
Unzipped binary to C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq/neestimator</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co">#simulate a population of 50 individuals with 1000 loci</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">50</span>, <span class="at">nlocs =</span> <span class="dv">3000</span>)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a><span class="co">#need to reproduce a bit (to get the population to be in Hardy-Weinberg equilibrium)</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>) <span class="co">#ideal population</span></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>) <span class="co">#ideal population</span></span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>) <span class="co">#ideal population</span></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>) <span class="co">#ideal population</span></span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>) <span class="co">#ideal population</span></span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span><span class="fu">gl.LDNe</span>(sim50, <span class="at">neest.path =</span> dir, <span class="at">mating =</span> <span class="st">"random"</span>, <span class="at">critical =</span> <span class="fu">c</span>(<span class="fl">0.1</span>,<span class="fl">0.05</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.LDNe 
  Processing genlight object with SNP data
Starting gl2genepop 
  Processing genlight object with SNP data
  The genepop file is saved as:  C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq/dummy.gen/
Completed: gl2genepop 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$offspring_Po
                    Statistic Frequency 1 Frequency 2 Frequency 3 Frequency 4
 Lowest Allele Frequency Used       0.100       0.050       No S*          0+
    Harmonic Mean Sample Size          50          50          50          50
      Independent Comparisons      662976     1195831     1725153     1947351
                  OverAll r^2    0.027755    0.027571    0.026941     0.02643
          Expected r^2 Sample    0.021276    0.021276    0.021276    0.021276
                Estimated Ne^        49.3        50.8        56.7        62.5
            CI low Parametric        48.5        50.2        56.1        61.9
           CI high Parametric          50        51.4        57.3        63.2
             CI low JackKnife        37.7        39.6        45.1        50.3
            CI high JackKnife        67.6        67.9          74        80.6

  The results are saved in: C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq/genepopLD.txt 
Completed: gl.LDNe </code></pre>
</div>
</div>
</section>
<section id="historic-population-sizes" class="level3">
<h3 class="anchored" data-anchor-id="historic-population-sizes">Historic population sizes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install the Neestimator package</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>dir <span class="ot">&lt;-</span> dartRverse<span class="sc">::</span><span class="fu">gl.download.binary</span>(<span class="st">"epos"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Downloaded binary to  C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq\file124457e62304 
Unzipped binary to C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq/epos</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.sfs</span>(sim50)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.sfs 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.sfs </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  d0   d1   d2   d3   d4   d5   d6   d7   d8   d9  d10  d11  d12  d13  d14  d15 
1026  116  129  111   71   98   72   85   70   70   64   54   57   49   34   37 
 d16  d17  d18  d19  d20  d21  d22  d23  d24  d25  d26  d27  d28  d29  d30  d31 
  38   37   34   25   30   37   24   22   16   18   30   19   22   30   26   29 
 d32  d33  d34  d35  d36  d37  d38  d39  d40  d41  d42  d43  d44  d45  d46  d47 
  34   27   22    9   21   18   17   17   26   25   26   25   17   30   17   34 
 d48  d49  d50 
  19   23   13 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">#parameters for epos</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="co"># mutation rate</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>u <span class="ot">=</span><span class="fl">1e-8</span>  <span class="co">#from simulation</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co">#To get L we need estimate the Length of the genome in base pairs.</span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co"># watterson estimate for a sample of 50 is 4.499</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>L <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">gl.sfs</span>(sim50)[<span class="sc">-</span><span class="dv">1</span>])<span class="sc">/</span>( <span class="dv">4</span> <span class="sc">*</span> <span class="dv">50</span> <span class="sc">*</span><span class="fl">1e-8</span> <span class="sc">*</span> <span class="fl">4.499</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.sfs 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.sfs </code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>gepos <span class="ot">&lt;-</span> <span class="fu">gl.run.epos</span>(sim50, <span class="at">epos.path =</span> dir, <span class="at">L=</span>L, <span class="at">u=</span>u, <span class="at">boot=</span><span class="dv">10</span>, <span class="at">minbinsize =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data
  Output written to C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq/epos.out </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-28-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.run.epos </code></pre>
</div>
</div>
<section id="cged-example" class="level4">
<h4 class="anchored" data-anchor-id="cged-example">CGED example</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>north <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/TympoNorth.rds"</span>)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="co">#filters </span></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>north2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(north, <span class="at">method =</span> <span class="st">"loc"</span>, <span class="at">threshold =</span> <span class="fl">0.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>north3 <span class="ot">&lt;-</span> <span class="fu">gl.filter.rdepth</span>(north2, <span class="at">lower =</span> <span class="dv">10</span>, <span class="at">upper=</span><span class="dv">40</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.rdepth 
  Processing genlight object with SNP data
  Removing loci with rdepth &lt;= 10 and &gt;= 40 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-29-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.rdepth </code></pre>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>north4 <span class="ot">&lt;-</span> <span class="fu">gl.filter.monomorphs</span>(north3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.monomorphs 
  Processing genlight object with SNP data
  Identifying monomorphic loci
  Removing monomorphic loci and loci with all missing 
                       data
Completed: gl.filter.monomorphs </code></pre>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>north5 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(north4, <span class="at">threshold =</span> <span class="fl">0.9</span>, <span class="at">method=</span><span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Recalculating Call Rate
  Removing individuals based on Call Rate, threshold = 0.9 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-29-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Locus metrics not recalculated
  Note: Resultant monomorphic loci not deleted
Completed: gl.filter.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co">#checks</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>north5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 121 genotypes,  10,832 SNPs , size: 44.6 Mb

    missing data: 13360 (=1.02 %) scored as NA

 ** Genetic data
   @gen: list of 121 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  121 individual labels
   @loc.names:  10832 locus labels
   @loc.all:  10832 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 121-121)
   @other: a list containing: loc.metrics, latlon, loc.metrics.flags, verbose, history 
   @other$latlon[g]: coordinates for all individuals are attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.sfs</span>(north5)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.sfs 
  Processing genlight object with SNP data
Your data contains missing data, better filter stronger or use gl.impute to fill those gaps meaningful!</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-29-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.sfs </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  d0   d1   d2   d3   d4   d5   d6   d7   d8   d9  d10  d11  d12  d13  d14  d15 
   0  747  557  375  244  203  165  154  118  125  122  114  114  122  107  115 
 d16  d17  d18  d19  d20  d21  d22  d23  d24  d25  d26  d27  d28  d29  d30  d31 
  96  106  102   96  104  106   79  105  103   98   78   89   86   79   91   89 
 d32  d33  d34  d35  d36  d37  d38  d39  d40  d41  d42  d43  d44  d45  d46  d47 
  86   89   83  101   95   78   83   87   53   90   77   89   78   76   73   90 
 d48  d49  d50  d51  d52  d53  d54  d55  d56  d57  d58  d59  d60  d61  d62  d63 
  66   75   88  100   75   70   71   80   75   72   77   54   81   74   63   67 
 d64  d65  d66  d67  d68  d69  d70  d71  d72  d73  d74  d75  d76  d77  d78  d79 
  65   90   64   63   70   49   72   74   60   61   72   60   50   70   62   71 
 d80  d81  d82  d83  d84  d85  d86  d87  d88  d89  d90  d91  d92  d93  d94  d95 
  49   70   70   64   69   58   62   51   76   51   63   54   59   56   44   61 
 d96  d97  d98  d99 d100 d101 d102 d103 d104 d105 d106 d107 d108 d109 d110 d111 
  51   61   54   54   76   42   49   62   43   49   56   37   53   53   53   62 
d112 d113 d114 d115 d116 d117 d118 d119 d120 d121 
  67   60   60   61   55   66   53   43   58   39 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nInd</span>(north5)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 121</code></pre>
</div>
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nLoc</span>(north5)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10832</code></pre>
</div>
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="do">### takes 10 minutes or so....</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="co">#curNe &lt;- gl.LDNe(north5, neest.path = dir, critical = 0.05)  #-&gt; ~40</span></span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a><span class="co">#for dart data (times 75), 69+</span></span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>L <span class="ot">&lt;-</span> <span class="fu">nLoc</span>(north5)<span class="sc">*</span><span class="dv">75</span><span class="sc">*</span><span class="dv">200</span></span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a>mu <span class="ot">=</span> <span class="fl">16.17e-9</span></span>
<span id="cb108-8"><a href="#cb108-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-9"><a href="#cb108-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-10"><a href="#cb108-10" aria-hidden="true" tabindex="-1"></a><span class="co"># takes a minute or so</span></span>
<span id="cb108-11"><a href="#cb108-11" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span> <span class="fu">gl.run.epos</span>(north5, <span class="at">epos.path =</span> dir,<span class="at">L =</span> L, <span class="at">u =</span> mu, </span>
<span id="cb108-12"><a href="#cb108-12" aria-hidden="true" tabindex="-1"></a>                  <span class="at">method =</span> <span class="st">"greedy"</span>, <span class="at">depth =</span> <span class="dv">2</span>, <span class="at">boot=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data
  Output written to C:\Users\ejstr\AppData\Local\Temp\RtmpyGf7Vq/epos.out </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-29-5.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.run.epos </code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>gg<span class="sc">$</span>plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session02_files/figure-html/unnamed-chunk-29-6.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="further-study" class="level2">
<h2 class="anchored" data-anchor-id="further-study"><em>Further Study</em></h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class img-fluid" width="48"> Now run your own data set (or use the tympo data set) to study any of the topics from the tutorial. Have fun !!!</p>
</div>
</div>
<section id="readings" class="level3">
<h3 class="anchored" data-anchor-id="readings">Readings</h3>
<ul>
<li><p>Campbell, C.D., Cowan, P., Gruber, B. et al.&nbsp;Has the introduction of two subspecies generated dispersal barriers among invasive possums in New Zealand?. Biol Invasions 23, 3831–3845 (2021). https://doi.org/10.1007/s10530-021-02609-1</p></li>
<li><p>Schmidt, T. L., Jasper, M.-E., Weeks, A. R., &amp; Hoffmann, A. A. (2021). Unbiased population heterozygosity estimates from genome-wide sequence data. Methods in Ecology and Evolution, 12, 1888–1898. https://doi.org/10.1111/2041-210X.13659</p></li>
<li><p>Sopniewski J, Catullo RA. Estimates of heterozygosity from single nucleotide polymorphism markers are context-dependent and often wrong. Mol Ecol Resour. 2024 May;24(4):e13947. doi: 10.1111/1755-0998.13947. Epub 2024 Mar 3. PMID: 38433491.</p></li>
<li><p>Wang, J. (2022). A joint likelihood estimator of relatedness and allele frequencies from a small sample of individuals. Methods in Ecology and Evolution, 13(11), 2443-2462.</p></li>
<li><p>Waples, R. S. (2006). “A bias correction for estimates of effective population size based on linkage disequilibrium at unlinked gene loci*.” Conservation Genetics 7(2): 167-184.</p></li>
<li><p>Waples, R. K., et al.&nbsp;(2016). “Estimating contemporary effective population size in non-model species using linkage disequilibrium across thousands of loci.” Heredity 117(4): 233-240.</p></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script>



/* update total correct if #webex-total_correct exists */

update_total_correct = function() {

  console.log("webex: update total_correct");



  var t = document.getElementsByClassName("webex-total_correct");

  for (var i = 0; i < t.length; i++) {

    p = t[i].parentElement;

    var correct = p.getElementsByClassName("webex-correct").length;

    var solvemes = p.getElementsByClassName("webex-solveme").length;

    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;

    var selects = p.getElementsByClassName("webex-select").length;



    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";

  }

}



/* webex-solution button toggling function */

b_func = function() {

  console.log("webex: toggle hide");



  var cl = this.parentElement.classList;

  if (cl.contains('open')) {

    cl.remove("open");

  } else {

    cl.add("open");

  }

}



/* check answers */

check_func = function() {

  console.log("webex: check answers");



  var cl = this.parentElement.classList;

  if (cl.contains('unchecked')) {

    cl.remove("unchecked");

    this.innerHTML = "Hide Answers";

  } else {

    cl.add("unchecked");

    this.innerHTML = "Show Answers";

  }

}



/* function for checking solveme answers */

solveme_func = function(e) {

  console.log("webex: check solveme");



  var real_answers = JSON.parse(this.dataset.answer);

  var my_answer = this.value;

  var cl = this.classList;

  if (cl.contains("ignorecase")) {

    my_answer = my_answer.toLowerCase();

  }

  if (cl.contains("nospaces")) {

    my_answer = my_answer.replace(/ /g, "")

  }



  if (my_answer == "") {

    cl.remove("webex-correct");

    cl.remove("webex-incorrect");

  } else if (real_answers.includes(my_answer)) {

    cl.add("webex-correct");

    cl.remove("webex-incorrect");

  } else {

    cl.add("webex-incorrect");

    cl.remove("webex-correct");

  }



  // match numeric answers within a specified tolerance

  if(this.dataset.tol > 0){

    var tol = JSON.parse(this.dataset.tol);

    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)

    if (matches.reduce((a, b) => a + b, 0) > 0) {

      cl.add("webex-correct");

    } else {

      cl.remove("webex-correct");

    }

  }



  // added regex bit

  if (cl.contains("regex")){

    answer_regex = RegExp(real_answers.join("|"))

    if (answer_regex.test(my_answer)) {

      cl.add("webex-correct");

    }

  }



  update_total_correct();

}



/* function for checking select answers */

select_func = function(e) {

  console.log("webex: check select");



  var cl = this.classList



  /* add style */

  cl.remove("webex-incorrect");

  cl.remove("webex-correct");

  if (this.value == "answer") {

    cl.add("webex-correct");

  } else if (this.value != "blank") {

    cl.add("webex-incorrect");

  }



  update_total_correct();

}



/* function for checking radiogroups answers */

radiogroups_func = function(e) {

  console.log("webex: check radiogroups");



  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');

  var cl = checked_button.parentElement.classList;

  var labels = checked_button.parentElement.parentElement.children;



  /* get rid of styles */

  for (i = 0; i < labels.length; i++) {

    labels[i].classList.remove("webex-incorrect");

    labels[i].classList.remove("webex-correct");

  }



  /* add style */

  if (checked_button.value == "answer") {

    cl.add("webex-correct");

  } else {

    cl.add("webex-incorrect");

  }



  update_total_correct();

}



window.onload = function() {

  console.log("webex onload");

  /* set up solution buttons */

  var buttons = document.getElementsByTagName("button");



  for (var i = 0; i < buttons.length; i++) {

    if (buttons[i].parentElement.classList.contains('webex-solution')) {

      buttons[i].onclick = b_func;

    }

  }



  var check_sections = document.getElementsByClassName("webex-check");

  console.log("check:", check_sections.length);

  for (var i = 0; i < check_sections.length; i++) {

    check_sections[i].classList.add("unchecked");



    let btn = document.createElement("button");

    btn.innerHTML = "Show Answers";

    btn.classList.add("webex-check-button");

    btn.onclick = check_func;

    check_sections[i].appendChild(btn);



    let spn = document.createElement("span");

    spn.classList.add("webex-total_correct");

    check_sections[i].appendChild(spn);

  }



  /* set up webex-solveme inputs */

  var solveme = document.getElementsByClassName("webex-solveme");



  for (var i = 0; i < solveme.length; i++) {

    /* make sure input boxes don't auto-anything */

    solveme[i].setAttribute("autocomplete","off");

    solveme[i].setAttribute("autocorrect", "off");

    solveme[i].setAttribute("autocapitalize", "off");

    solveme[i].setAttribute("spellcheck", "false");

    solveme[i].value = "";



    /* adjust answer for ignorecase or nospaces */

    var cl = solveme[i].classList;

    var real_answer = solveme[i].dataset.answer;

    if (cl.contains("ignorecase")) {

      real_answer = real_answer.toLowerCase();

    }

    if (cl.contains("nospaces")) {

      real_answer = real_answer.replace(/ /g, "");

    }

    solveme[i].dataset.answer = real_answer;



    /* attach checking function */

    solveme[i].onkeyup = solveme_func;

    solveme[i].onchange = solveme_func;



    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")

  }



  /* set up radiogroups */

  var radiogroups = document.getElementsByClassName("webex-radiogroup");

  for (var i = 0; i < radiogroups.length; i++) {

    radiogroups[i].onchange = radiogroups_func;

  }



  /* set up selects */

  var selects = document.getElementsByClassName("webex-select");

  for (var i = 0; i < selects.length; i++) {

    selects[i].onchange = select_func;

    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")

  }



  update_total_correct();

}



</script>

<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session01.html" class="pagination-link" aria-label="Importing and Handling Genomic Data">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Importing and Handling Genomic Data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session03.html" class="pagination-link" aria-label="Identifying Population Structure">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identifying Population Structure</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>