<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>5&nbsp; Assigning Individuals to Populations – Analysing Genomic Data with dartRverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session06.html" rel="next">
<link href="./session04.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-a185852c63625fd9ffbdc57047c9a77e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-ef31e3f4d9d6eef55c54bdf6b9cd68ca.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session05.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assigning Individuals to Populations</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Analysing Genomic Data with dartRverse</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./speakers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Speaker Bios</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./breakk.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Importing and Handling Genomic Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Estimating Effective Population Size and Key Stats</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Identifying Population Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Kinship, Pedigrees &amp; Mating Recommendations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assigning Individuals to Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">SNP Panel Selection &amp; Simulations for Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Simulations for Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Bonus! Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Session Materials</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./feedback.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Feedback</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#emydura-population-assignment" id="toc-emydura-population-assignment" class="nav-link" data-scroll-target="#emydura-population-assignment"><em>Emydura</em> population assignment</a>
  <ul>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#assignment-by-genotype-likelihood" id="toc-assignment-by-genotype-likelihood" class="nav-link" data-scroll-target="#assignment-by-genotype-likelihood">Assignment by genotype likelihood</a></li>
  <li><a href="#assignment-by-private-alleles" id="toc-assignment-by-private-alleles" class="nav-link" data-scroll-target="#assignment-by-private-alleles">Assignment by Private Alleles</a></li>
  <li><a href="#assignment-by-pca" id="toc-assignment-by-pca" class="nav-link" data-scroll-target="#assignment-by-pca">Assignment by PCA</a></li>
  <li><a href="#assignment-by-mahalanobis-distances" id="toc-assignment-by-mahalanobis-distances" class="nav-link" data-scroll-target="#assignment-by-mahalanobis-distances">Assignment by Mahalanobis Distances</a></li>
  </ul></li>
  <li><a href="#scenario" id="toc-scenario" class="nav-link" data-scroll-target="#scenario">Scenario</a>
  <ul>
  <li><a href="#the-data" id="toc-the-data" class="nav-link" data-scroll-target="#the-data">The data</a></li>
  </ul></li>
  <li><a href="#further-study" id="toc-further-study" class="nav-link" data-scroll-target="#further-study"><em>Further Study</em></a>
  <ul>
  <li><a href="#readings" id="toc-readings" class="nav-link" data-scroll-target="#readings">Readings</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Assigning Individuals to Populations</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><em>Session Presenter</em></p>
<p><img src="images/Presenters5.png" class="img-fluid"></p>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/turtle.jpg" class="img-fluid figure-img"></p>
<figcaption><em>Emydura</em> River Turtle</figcaption>
</figure>
</div>
<p>We will explore four analysis for assignment of an individual of unknown provenance to a source population.</p>
<ul>
<li><p><strong>Genotype Likelihood</strong>: The likelihood of drawing the unknown from a population with the observed allele frequencies is calculated assuming Hardy-Weinberg equilibrium.</p></li>
<li><p><strong>Private Alleles</strong>: A focal unknown individual is likely to have fewer private alleles in comparison with its source population than in comparison with other putative source populations.</p></li>
<li><p><strong>PCA</strong>: The genotype of a focal unknown individual is likely to lie within the confidence envelope of its source population than within the confidence envelope of other putative source populations.</p></li>
<li><p><strong>Mahalanobis Distance</strong>: The distances of the focal unknown individual from the centroids of the standardized confidence envelops of its putative source populations are used to calculate a z-scores and associated probabilities of assignment.</p></li>
</ul>
</section>
<section id="emydura-population-assignment" class="level2">
<h2 class="anchored" data-anchor-id="emydura-population-assignment"><em>Emydura</em> population assignment</h2>
<p>Here is some context for our turtle example, below are two maps. The first is a map of eastern mainland Australia showing the distribution of the <em>Emydura macquarii</em> complex, the river drainage basins in which it occurs, and the broad regions, The second is of Northern Australia and Papua New Guinea.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/turtle_map1.jpg" class="img-fluid figure-img"></p>
<figcaption><a href="https://onlinelibrary.wiley.com/doi/10.1111/mec.14925">Georges <em>et al.</em> 2018</a></figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/turtle_map2.jpg" class="img-fluid figure-img"></p>
<figcaption>Georges <em>et al.</em> 2025, in review</figcaption>
</figure>
</div>
<section id="load-data" class="level3">
<h3 class="anchored" data-anchor-id="load-data">Load data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will first set the verbosity globally to level 3</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.set.verbosity</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.set.verbosity 
  Global verbosity set to: 3 
Completed: gl.set.verbosity </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the data set for the worked example</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>gl <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/assignment_example1.Rdata"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Familiarize yourself with its contents</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>gl  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 835 genotypes,  20,688 SNPs , size: 57.3 Mb

    missing data: 289548 (=1.68 %) scored as NA

 ** Genetic data
   @gen: list of 835 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  835 individual labels
   @loc.names:  20688 locus labels
   @loc.all:  20688 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 3-30)
   @other: a list containing: loc.metrics, ind.metrics, latlon, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, lat, lon, sex, maturity, collector, location, basin, drainage, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC, TrimmedSequence 
   @other$latlon[g]: coordinates for all individuals are attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nLoc</span>(gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 20688</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nInd</span>(gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 835</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nPop</span>(gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 81</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Display a list of populations and sample sizes</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">pop</span>(gl))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
         Brisbane          Burdekin           Burnett          Clarence 
               10                10                11                10 
     Cooper_Alvin      Cooper_Cully  Cooper_Eulbertie        Dumaresque 
               10                10                10                10 
Fitzroy_Alligator  Fitzroy_Carnavan  Fitzroy_Fairburn     Fraser_Island 
               10                10                10                10 
           Hunter     EmmacJohnWari     EmmacMaclGeor              Mary 
               10                10                11                10 
     EmmacMDBBarr      EmmacMDBBarw     EmmacMDBBooth      EmmacMDBBowm 
               10                10                 9                10 
     EmmacMDBBurr      EmmacMDBCond      EmmacMDBCudg  EmmacMDBDarlBour 
               10                10                10                10 
 EmmacMDBDarlWeth      EmmacMDBDart      EmmacMDBEulo      EmmacMDBForb 
               10                10                10                10 
     EmmacMDBGoul        GurraGurra      EmmacMDBGwyd      EmmacMDBLach 
               10                10                10                10 
     EmmacMDBLodd      EmmacMDBMaci      EmmacMDBMoon  EmmacMDBMurrGunb 
               10                10                10                10 
 EmmacMDBMurrLock  EmmacMDBMurrMorg  EmmacMDBMurrMung  EmmacMDBMurrMurr 
               10                10                10                10 
 EmmacMDBMurrTink EmmacMDBMurrYarra      EmmacMDBOven  EmmacMDBParoBiny 
               10                10                10                10 
     EmmacMDBPind      EmmacMDBSanf      EmmacMDBToon          Normanby 
               10                10                11                11 
             Pine     EmmacRichCasi         EmmacRoss      EmmacTweeUki 
               10                10                10                10 
     EmsubBamuAli     EmsubBamuAwab     EmsubMorehead      EmsubFlyGuka 
               10                 9                16                10 
     EmsubFlyJikw      EmsubJardine       EmsubKerema       EmsubKikori 
               30                16                10                 4 
       EmworRoper        EmtanBlyth      EmtanFinniss     EmtanHolrChai 
               11                10                 7                10 
    EmtanMitchell     EmtanMitcMitc     EmtanPascFarm      EmtanWenlock 
                9                 3                 9                10 
        EmvicDaly     EmvicDrysdale        Fitzroy_WA     EmvicIsdeBell 
               10                10                10                12 
    EmvicKingMool          EmvicOrd     EmworClavPung         EmworDaly 
               10                18                10                10 
    EmworDalySlei     EmworLeicAlex     EmworLimmNath     EmworLiveMann 
                7                10                10                 9 
    EmworNichGreg 
               12 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Several populations have sample sizes less than 10 and will be discarded during the analysis</p>
</div>
</div>
</section>
<section id="assignment-by-genotype-likelihood" class="level3">
<h3 class="anchored" data-anchor-id="assignment-by-genotype-likelihood">Assignment by genotype likelihood</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gen.result<span class="ot">&lt;-</span><span class="fu">gl.assign.on.genotype</span>(gl, <span class="at">unknown=</span><span class="st">"AA011731"</span>, <span class="at">nmin=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.assign.on.genotype 
  Processing genlight object with SNP data
          population Log Likelihood        AIC        dAIC        AIC.wt assign
3            Burnett      -4926.957   9853.914      0.0000  1.000000e+00    yes
16              Mary      -5341.050  10682.101    828.1863 1.450906e-180     no
1           Brisbane     -19251.444  38502.888  28648.9733  0.000000e+00     no
2           Burdekin     -32844.476  65688.953  55835.0384  0.000000e+00     no
4           Clarence     -31620.048  63240.095  53386.1808  0.000000e+00     no
5       Cooper_Alvin     -42008.293  84016.586  74162.6716  0.000000e+00     no
6       Cooper_Cully     -42849.639  85699.278  75845.3633  0.000000e+00     no
7   Cooper_Eulbertie     -42636.382  85272.764  75418.8497  0.000000e+00     no
8         Dumaresque     -28852.254  57704.509  47850.5946  0.000000e+00     no
9  Fitzroy_Alligator     -12133.240  24266.480  14412.5655  0.000000e+00     no
10  Fitzroy_Carnavan     -13118.904  26237.808  16383.8939  0.000000e+00     no
11  Fitzroy_Fairburn     -12281.682  24563.364  14709.4500  0.000000e+00     no
12     Fraser_Island     -17364.115  34728.231  24874.3162  0.000000e+00     no
13            Hunter     -49665.375  99330.751  89476.8365  0.000000e+00     no
14     EmmacJohnWari     -35711.561  71423.121  61569.2070  0.000000e+00     no
15     EmmacMaclGeor     -37389.885  74779.769  64925.8549  0.000000e+00     no
17      EmmacMDBBarr     -29473.892  58947.783  49093.8690  0.000000e+00     no
18      EmmacMDBBarw     -29276.257  58552.514  48698.5994  0.000000e+00     no
19      EmmacMDBBowm     -30859.543  61719.085  51865.1709  0.000000e+00     no
20      EmmacMDBBurr     -32296.475  64592.951  54739.0361  0.000000e+00     no
21      EmmacMDBCond     -25487.072  50974.145  41120.2302  0.000000e+00     no
22      EmmacMDBCudg     -29695.945  59391.891  49537.9762  0.000000e+00     no
23  EmmacMDBDarlBour     -28964.950  57929.900  48075.9852  0.000000e+00     no
24  EmmacMDBDarlWeth     -27302.477  54604.954  44751.0400  0.000000e+00     no
25      EmmacMDBDart     -33067.630  66135.261  56281.3466  0.000000e+00     no
26      EmmacMDBEulo     -20835.127  41670.254  31816.3394  0.000000e+00     no
27      EmmacMDBForb     -32680.705  65361.409  55507.4948  0.000000e+00     no
28      EmmacMDBGoul     -29137.664  58275.328  48421.4131  0.000000e+00     no
29        GurraGurra     -29715.298  59430.595  49576.6806  0.000000e+00     no
30      EmmacMDBGwyd     -29369.851  58739.701  48885.7868  0.000000e+00     no
31      EmmacMDBLach     -32476.968  64953.935  55100.0211  0.000000e+00     no
32      EmmacMDBLodd     -29746.861  59493.723  49639.8084  0.000000e+00     no
33      EmmacMDBMaci     -28944.433  57888.866  48034.9512  0.000000e+00     no
34      EmmacMDBMoon     -29257.025  58514.050  48660.1356  0.000000e+00     no
35  EmmacMDBMurrGunb     -28266.313  56532.626  46678.7116  0.000000e+00     no
36  EmmacMDBMurrLock     -29738.313  59476.625  49622.7109  0.000000e+00     no
37  EmmacMDBMurrMorg     -29058.787  58117.573  48263.6589  0.000000e+00     no
38  EmmacMDBMurrMung     -29471.357  58942.715  49088.8005  0.000000e+00     no
39  EmmacMDBMurrMurr     -29727.568  59455.137  49601.2226  0.000000e+00     no
40  EmmacMDBMurrTink     -28706.653  57413.305  47559.3909  0.000000e+00     no
41 EmmacMDBMurrYarra     -29584.123  59168.246  49314.3315  0.000000e+00     no
42      EmmacMDBOven     -29769.909  59539.819  49685.9045  0.000000e+00     no
43  EmmacMDBParoBiny     -30378.843  60757.686  50903.7713  0.000000e+00     no
44      EmmacMDBPind     -32423.222  64846.444  54992.5292  0.000000e+00     no
45      EmmacMDBSanf     -30946.657  61893.315  52039.4005  0.000000e+00     no
46      EmmacMDBToon     -21983.158  43966.316  34112.4020  0.000000e+00     no
47          Normanby     -42201.524  84403.048  74549.1332  0.000000e+00     no
48              Pine      -8578.105  17156.211   7302.2964  0.000000e+00     no
49     EmmacRichCasi     -25491.051  50982.103  41128.1884  0.000000e+00     no
50         EmmacRoss     -34003.894  68007.788  58153.8734  0.000000e+00     no
51      EmmacTweeUki     -20216.624  40433.249  30579.3343  0.000000e+00     no
52      EmsubBamuAli     -84564.461 169128.921 159275.0068  0.000000e+00     no
53     EmsubMorehead     -81201.565 162403.130 152549.2157  0.000000e+00     no
54      EmsubFlyGuka     -83272.339 166544.677 156690.7630  0.000000e+00     no
55      EmsubFlyJikw     -79486.100 158972.200 149118.2860  0.000000e+00     no
56      EmsubJardine     -86359.491 172718.983 162865.0685  0.000000e+00     no
57       EmsubKerema     -94433.639 188867.278 179013.3636  0.000000e+00     no
58        EmworRoper     -84920.595 169841.191 159987.2763  0.000000e+00     no
59        EmtanBlyth    -103876.148 207752.297 197898.3826  0.000000e+00     no
60     EmtanHolrChai    -101420.838 202841.676 192987.7618  0.000000e+00     no
61      EmtanWenlock    -101328.752 202657.504 192803.5897  0.000000e+00     no
62         EmvicDaly     -85769.402 171538.805 161684.8903  0.000000e+00     no
63     EmvicDrysdale     -97248.693 194497.386 184643.4716  0.000000e+00     no
64        Fitzroy_WA     -96787.705 193575.409 183721.4946  0.000000e+00     no
65     EmvicIsdeBell     -95745.919 191491.838 181637.9234  0.000000e+00     no
66     EmvicKingMool     -96702.092 193404.185 183550.2704  0.000000e+00     no
67          EmvicOrd     -93520.359 187040.717 177186.8030  0.000000e+00     no
68     EmworClavPung     -88270.073 176540.145 166686.2306  0.000000e+00     no
69         EmworDaly     -90298.465 180596.931 170743.0165  0.000000e+00     no
70     EmworLeicAlex     -91795.194 183590.389 173736.4745  0.000000e+00     no
71     EmworLimmNath     -91979.158 183958.316 174104.4020  0.000000e+00     no
72     EmworNichGreg     -85210.317 170420.635 160566.7205  0.000000e+00     no
  Warning: parameter by must be either 'join.by.ind' or 'join.by.loc', set to default 'join.by.loc'
Completed: gl.assign.on.genotype </code></pre>
</div>
</div>
</section>
<section id="assignment-by-private-alleles" class="level3">
<h3 class="anchored" data-anchor-id="assignment-by-private-alleles">Assignment by Private Alleles</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pa.result <span class="ot">&lt;-</span> <span class="fu">gl.assign.pa</span>(gl, <span class="at">unknown=</span><span class="st">"AA011731"</span>, <span class="at">nmin=</span><span class="dv">10</span>, <span class="at">alpha=</span><span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.assign.pa 
  Processing genlight object with SNP data
  Discarding 9 populations with sample size &lt; 10 :
                 pop count    Z-score  p-value assign
16              Mary    81 -0.1692350 0.567194    yes
3            Burnett    77  0.2743299 0.391916    yes
49              Pine   167  1.1555039 0.123942    yes
22      EmmacMDBCond   785  2.0204271 0.021670     no
47      EmmacMDBToon   668  2.7347470 0.003121     no
15     EmmacMaclGeor  1040  3.4791497 0.000252     no
69         EmvicDaly  1284  3.5437788 0.000197     no
20      EmmacMDBBowm   992  3.6051586 0.000156     no
81     EmworNichGreg  1260  3.8784997 0.000053     no
61        EmworRoper  1273  4.1008215 0.000021     no
25  EmmacMDBDarlWeth   865  4.8762430 0.000001     no
1           Brisbane   523 17.1445337 0.000000     no
2           Burdekin   821 15.7683126 0.000000     no
4           Clarence   915 14.1798931 0.000000     no
5       Cooper_Alvin   992 15.6406240 0.000000     no
6       Cooper_Cully  1008 18.1191005 0.000000     no
7   Cooper_Eulbertie  1001  9.1930282 0.000000     no
8         Dumaresque   929 31.3868949 0.000000     no
9  Fitzroy_Alligator   306 11.4289937 0.000000     no
10  Fitzroy_Carnavan   339 10.2951686 0.000000     no
11  Fitzroy_Fairburn   303  7.6467045 0.000000     no
12     Fraser_Island   457  5.0064421 0.000000     no
13            Hunter  1340 12.4578521 0.000000     no
14     EmmacJohnWari   893 11.1739703 0.000000     no
17      EmmacMDBBarr   940 26.7349588 0.000000     no
18      EmmacMDBBarw   937 26.1751262 0.000000     no
19     EmmacMDBBooth  1035 15.7478733 0.000000     no
21      EmmacMDBBurr  1025 15.2447327 0.000000     no
23      EmmacMDBCudg   952 18.4925912 0.000000     no
24  EmmacMDBDarlBour   916 13.1037035 0.000000     no
26      EmmacMDBDart  1079 14.9852921 0.000000     no
27      EmmacMDBEulo   639  6.1130987 0.000000     no
28      EmmacMDBForb  1051  5.2168208 0.000000     no
29      EmmacMDBGoul   922 12.4532508 0.000000     no
30        GurraGurra   957 13.0155533 0.000000     no
31      EmmacMDBGwyd   940 23.3109909 0.000000     no
32      EmmacMDBLach  1053 17.7866486 0.000000     no
33      EmmacMDBLodd   950 16.0172441 0.000000     no
34      EmmacMDBMaci   925 15.1899478 0.000000     no
35      EmmacMDBMoon   928 21.5894040 0.000000     no
36  EmmacMDBMurrGunb   898  7.6464411 0.000000     no
37  EmmacMDBMurrLock   959  8.1186128 0.000000     no
38  EmmacMDBMurrMorg   922 15.3803498 0.000000     no
39  EmmacMDBMurrMung   946 15.3622303 0.000000     no
40  EmmacMDBMurrMurr   958 27.7218281 0.000000     no
41  EmmacMDBMurrTink   912 11.1714406 0.000000     no
42 EmmacMDBMurrYarra   950 27.2732611 0.000000     no
43      EmmacMDBOven   949 27.2094137 0.000000     no
44  EmmacMDBParoBiny   975 11.8093091 0.000000     no
45      EmmacMDBPind  1037 25.1472989 0.000000     no
46      EmmacMDBSanf   995 20.6254532 0.000000     no
48          Normanby  1014  8.9353965 0.000000     no
50     EmmacRichCasi   727 23.4264098 0.000000     no
51         EmmacRoss   853 16.4775772 0.000000     no
52      EmmacTweeUki   591 10.7132631 0.000000     no
53      EmsubBamuAli  1286 23.2269725 0.000000     no
54     EmsubBamuAwab  1285 16.8725666 0.000000     no
55     EmsubMorehead  1238 21.7595831 0.000000     no
56      EmsubFlyGuka  1268 25.9689306 0.000000     no
57      EmsubFlyJikw  1226 17.9295179 0.000000     no
58      EmsubJardine  1287 10.9701686 0.000000     no
59       EmsubKerema  1370  5.1264437 0.000000     no
60       EmsubKikori  1326 20.7904053 0.000000     no
62        EmtanBlyth  1396  8.9673202 0.000000     no
63      EmtanFinniss  1382  7.6892448 0.000000     no
64     EmtanHolrChai  1361 10.7419476 0.000000     no
65     EmtanMitchell  1343  5.4248309 0.000000     no
66     EmtanMitcMitc  1369 12.8371209 0.000000     no
67     EmtanPascFarm  1365 17.4254784 0.000000     no
68      EmtanWenlock  1351 10.6423539 0.000000     no
70     EmvicDrysdale  1365 13.7459176 0.000000     no
71        Fitzroy_WA  1372 10.2841962 0.000000     no
72     EmvicIsdeBell  1355 14.7314585 0.000000     no
73     EmvicKingMool  1363 24.4944007 0.000000     no
74          EmvicOrd  1333 12.5867638 0.000000     no
75     EmworClavPung  1299 22.5017244 0.000000     no
76         EmworDaly  1307  5.2935238 0.000000     no
77     EmworDalySlei  1321  4.8949185 0.000000     no
78     EmworLeicAlex  1324 15.9637009 0.000000     no
79     EmworLimmNath  1322  5.7857267 0.000000     no
80     EmworLiveMann  1331 19.8407465 0.000000     no
  Warning: parameter by must be either 'join.by.ind' or 'join.by.loc', set to default 'join.by.loc'
Completed: gl.assign.pa </code></pre>
</div>
</div>
</section>
<section id="assignment-by-pca" class="level3">
<h3 class="anchored" data-anchor-id="assignment-by-pca">Assignment by PCA</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pca_pa_result <span class="ot">&lt;-</span><span class="fu">gl.assign.pca</span>(pa.result, <span class="at">unknown=</span><span class="st">"AA011731"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.assign.pca 
Starting gl.keep.pop 
  Processing genlight object with SNP data
  Checking for presence of nominated populations
  Retaining only populations Unknown 
  Locus metrics not recalculated
Completed: gl.keep.pop 
  Discarding 0 populations with sample size &lt; nmin = 10 :</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Calculating a PCA to represent the unknown in the context
                   of putative sources</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session05_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session05_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Eliminating populations for which the unknown is outside
                   their confidence envelope
  Returning a genlight object with remaining putative source
                   populations plus the unknown
Completed: gl.assign.pca </code></pre>
</div>
</div>
</section>
<section id="assignment-by-mahalanobis-distances" class="level3">
<h3 class="anchored" data-anchor-id="assignment-by-mahalanobis-distances">Assignment by Mahalanobis Distances</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>mahal_result <span class="ot">&lt;-</span> <span class="fu">gl.assign.mahal</span>(pa.result,<span class="at">unknown=</span><span class="st">"AA011731"</span>, <span class="at">verbose =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>HELLO HELLO HELLO
Starting gl.assign.mahal 
  Discarding 0 populations with sample size &lt; 10 :
 
  Rendering the data matrix dense by imputation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Undertaking a PCA</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="session05_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
  Dimensions retained: 1 
  Number of dimensions with substantial eigenvalues (Broken-Stick Criterion): 1 . Hardwired limit 3 
    Selecting the smallest of the two
    Dimension of confidence envelope set at 1 
Assignment of unknown individual: AA011731 
Alpha level of significance: 0.001 
      pop     MahalD      pval assign
1    Mary 0.09623112 1.0000000    yes
2    Pine 0.34572127 0.9999989    yes
3 Burnett 3.32087394 0.9728334    yes
  Best assignment is the population with the largest probability
                of assignment, in this case Mary 
  Returning a genlight object with the putative source populations and the unknown
Completed: gl.assign.mahal </code></pre>
</div>
</div>
</section>
</section>
<section id="scenario" class="level2">
<h2 class="anchored" data-anchor-id="scenario">Scenario</h2>
<p>The authorities have recently raided a premises in Brisbane and found a number of reptiles held without permit. One of these is the painted turtle <strong>Emydura subglobosa</strong>. This species is widespread and common in southern New Guinea, but restricted in Australia to the Jardine River at the tip of Cape York. The Australian population is considered critically endangered under the EPBC Act. The question is, was the animal sourced from Cape York or imported from New Guinea? The specimen was genotyped and run in a service with the other available specimens from localities shown in Figure 1. The datafile is <code>assignment_example1.Rdata</code>. The SpecimenID is “AA046092“. Before you begin the analysis, restrict the populations under consideration to Emydura subglobosa.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class img-fluid" width="48"> Can you confidently decide if the animal was sourced from Cape York or New Guinea using the tools we have provided you via dartR?</p>
</div>
</div>
<section id="the-data" class="level3">
<h3 class="anchored" data-anchor-id="the-data">The data</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>gl</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="co"># The unknown</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>Unknown <span class="ot">=</span> <span class="st">"AA046092"</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Preliminaries</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">popNames</span>(gl)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>gl2 <span class="ot">&lt;-</span> <span class="fu">gl.keep.pop</span>(gl, <span class="at">pop.list=</span><span class="fu">c</span>(<span class="st">"EmsubBamuAli"</span>, <span class="st">"EmsubFlyGuka"</span>, <span class="st">"EmsubFlyJikw"</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"EmsubJardine"</span>, <span class="st">"EmsubKerema"</span>, <span class="st">"EmsubMorehead"</span>))</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Knock yourself out</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="further-study" class="level2">
<h2 class="anchored" data-anchor-id="further-study"><em>Further Study</em></h2>
<p>Tutorial yet to come…</p>
<section id="readings" class="level3">
<h3 class="anchored" data-anchor-id="readings">Readings</h3>


</section>
</section>

</main> <!-- /main -->
<script>



/* update total correct if #webex-total_correct exists */

update_total_correct = function() {

  console.log("webex: update total_correct");



  var t = document.getElementsByClassName("webex-total_correct");

  for (var i = 0; i < t.length; i++) {

    p = t[i].parentElement;

    var correct = p.getElementsByClassName("webex-correct").length;

    var solvemes = p.getElementsByClassName("webex-solveme").length;

    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;

    var selects = p.getElementsByClassName("webex-select").length;



    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";

  }

}



/* webex-solution button toggling function */

b_func = function() {

  console.log("webex: toggle hide");



  var cl = this.parentElement.classList;

  if (cl.contains('open')) {

    cl.remove("open");

  } else {

    cl.add("open");

  }

}



/* check answers */

check_func = function() {

  console.log("webex: check answers");



  var cl = this.parentElement.classList;

  if (cl.contains('unchecked')) {

    cl.remove("unchecked");

    this.innerHTML = "Hide Answers";

  } else {

    cl.add("unchecked");

    this.innerHTML = "Show Answers";

  }

}



/* function for checking solveme answers */

solveme_func = function(e) {

  console.log("webex: check solveme");



  var real_answers = JSON.parse(this.dataset.answer);

  var my_answer = this.value;

  var cl = this.classList;

  if (cl.contains("ignorecase")) {

    my_answer = my_answer.toLowerCase();

  }

  if (cl.contains("nospaces")) {

    my_answer = my_answer.replace(/ /g, "")

  }



  if (my_answer == "") {

    cl.remove("webex-correct");

    cl.remove("webex-incorrect");

  } else if (real_answers.includes(my_answer)) {

    cl.add("webex-correct");

    cl.remove("webex-incorrect");

  } else {

    cl.add("webex-incorrect");

    cl.remove("webex-correct");

  }



  // match numeric answers within a specified tolerance

  if(this.dataset.tol > 0){

    var tol = JSON.parse(this.dataset.tol);

    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)

    if (matches.reduce((a, b) => a + b, 0) > 0) {

      cl.add("webex-correct");

    } else {

      cl.remove("webex-correct");

    }

  }



  // added regex bit

  if (cl.contains("regex")){

    answer_regex = RegExp(real_answers.join("|"))

    if (answer_regex.test(my_answer)) {

      cl.add("webex-correct");

    }

  }



  update_total_correct();

}



/* function for checking select answers */

select_func = function(e) {

  console.log("webex: check select");



  var cl = this.classList



  /* add style */

  cl.remove("webex-incorrect");

  cl.remove("webex-correct");

  if (this.value == "answer") {

    cl.add("webex-correct");

  } else if (this.value != "blank") {

    cl.add("webex-incorrect");

  }



  update_total_correct();

}



/* function for checking radiogroups answers */

radiogroups_func = function(e) {

  console.log("webex: check radiogroups");



  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');

  var cl = checked_button.parentElement.classList;

  var labels = checked_button.parentElement.parentElement.children;



  /* get rid of styles */

  for (i = 0; i < labels.length; i++) {

    labels[i].classList.remove("webex-incorrect");

    labels[i].classList.remove("webex-correct");

  }



  /* add style */

  if (checked_button.value == "answer") {

    cl.add("webex-correct");

  } else {

    cl.add("webex-incorrect");

  }



  update_total_correct();

}



window.onload = function() {

  console.log("webex onload");

  /* set up solution buttons */

  var buttons = document.getElementsByTagName("button");



  for (var i = 0; i < buttons.length; i++) {

    if (buttons[i].parentElement.classList.contains('webex-solution')) {

      buttons[i].onclick = b_func;

    }

  }



  var check_sections = document.getElementsByClassName("webex-check");

  console.log("check:", check_sections.length);

  for (var i = 0; i < check_sections.length; i++) {

    check_sections[i].classList.add("unchecked");



    let btn = document.createElement("button");

    btn.innerHTML = "Show Answers";

    btn.classList.add("webex-check-button");

    btn.onclick = check_func;

    check_sections[i].appendChild(btn);



    let spn = document.createElement("span");

    spn.classList.add("webex-total_correct");

    check_sections[i].appendChild(spn);

  }



  /* set up webex-solveme inputs */

  var solveme = document.getElementsByClassName("webex-solveme");



  for (var i = 0; i < solveme.length; i++) {

    /* make sure input boxes don't auto-anything */

    solveme[i].setAttribute("autocomplete","off");

    solveme[i].setAttribute("autocorrect", "off");

    solveme[i].setAttribute("autocapitalize", "off");

    solveme[i].setAttribute("spellcheck", "false");

    solveme[i].value = "";



    /* adjust answer for ignorecase or nospaces */

    var cl = solveme[i].classList;

    var real_answer = solveme[i].dataset.answer;

    if (cl.contains("ignorecase")) {

      real_answer = real_answer.toLowerCase();

    }

    if (cl.contains("nospaces")) {

      real_answer = real_answer.replace(/ /g, "");

    }

    solveme[i].dataset.answer = real_answer;



    /* attach checking function */

    solveme[i].onkeyup = solveme_func;

    solveme[i].onchange = solveme_func;



    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")

  }



  /* set up radiogroups */

  var radiogroups = document.getElementsByClassName("webex-radiogroup");

  for (var i = 0; i < radiogroups.length; i++) {

    radiogroups[i].onchange = radiogroups_func;

  }



  /* set up selects */

  var selects = document.getElementsByClassName("webex-select");

  for (var i = 0; i < selects.length; i++) {

    selects[i].onchange = select_func;

    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")

  }



  update_total_correct();

}



</script>

<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session04.html" class="pagination-link" aria-label="Kinship, Pedigrees &amp; Mating Recommendations">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Kinship, Pedigrees &amp; Mating Recommendations</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session06.html" class="pagination-link" aria-label="SNP Panel Selection &amp; Simulations for Conservation">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">SNP Panel Selection &amp; Simulations for Conservation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>